minvws/nl-kat-coordination
# Add s3 functionality in Bytes

### Changes

Made it so that if the user adds `S3_BUCKET_PREFIX` or `S3_BUCKET` to their environment variables when building OpenKAT, that OpenKAT will use S3-buckets for saving the data from boefje's raw files instead of saving them on the local disk.


### Issue link

Closes #3191 

### Demo

Example files made inside s3
![image](https://github.com/user-attachments/assets/091d9220-abf8-422e-8966-d5f8c4900881)

Example of one of the files
![image](https://github.com/user-attachments/assets/6b8dcbd2-dc6a-4888-9a72-fea828846934)


### QA notes

To test this feature you will need access to an existing s3 server. And inside of you `.env` file you will need to provide the following variables which are required for [boto3](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#using-environment-variables):

- AWS_ACCESS_KEY_ID
- AWS_SECRET_ACCESS_KEY
- AWS_ENDPOINT_URL

Outside of these you also have to specify either of the following variables to let OpenKAT know you want to use s3 for the storage of bytes:

- S3_BUCKET_PREFIX
- S3_BUCKET

Make sure that these are compliant with the [bucket naming rules of AWS](https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html).

These are the variables I used for testing:
```env
S3_BUCKET_PREFIX=souf-prefix
S3_BUCKET=souf-bucket

# Required env variables for `boto3`
AWS_ACCESS_KEY_ID=GKbf4e89d08d6c5250e7f1de60
AWS_SECRET_ACCESS_KEY=406fef311a69f02fa1c3b58d73a6299651c7890dc0ef5448104f1762e29887b7
AWS_ENDPOINT_URL=http://garage:3900
```

I made use of the s3 service [garage](https://garagehq.deuxfleurs.fr/)
[How to setup garage](https://garagehq.deuxfleurs.fr/documentation/quick-start/)
[How to work with garage in a container](https://garagehq.deuxfleurs.fr/documentation/cookbook/real-world/)

I built a container for garage with the following docker compose file on the same host machine as `bytes`
```yml
services:
  garage:
    image: dxflrs/garage:v1.0.0
    restart: unless-stopped
    ports:
      - "127.0.0.1:3900:3900"
    volumes:
      - /etc/garage.toml:/etc/garage.toml
      - /var/lib/garage/meta:/var/lib/garage/meta
      - /var/lib/garage/data:/var/lib/garage/data
    networks:
      - network1

networks:
  network1:
    name: nl-kat-coordination_default
    external: true
```

---

### Code Checklist







---

## Checklist for code reviewers:


---

## Checklist for QA:


### What works:

- _bullet point + screenshot (if useful) per tested functionality_

### What doesn't work:

- _bullet point + screenshot (if useful) per tested functionality_

### Bug or feature?:

- _bullet point + screenshot (if useful) if it is unclear whether something is a bug or an intended feature._
